op-program/client/l2/canon_test.go

```diff
@@ -0,0 +1,108 @@
+package l2
+
+import (
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCanonicalBlockNumberOracle_GetHeaderByNumber(t *testing.T) {
+	headBlockNumber := 3
+	blockCount := 3
+	chainCfg, blocks, oracle := setupOracle(t, blockCount, headBlockNumber, true)
+	head := blocks[headBlockNumber].Header()
+
+	blockByHash := func(hash common.Hash) *types.Block {
+		return oracle.BlockByHash(hash, chainCfg.ChainID.Uint64())
+	}
```

Comment:
We should ensure we don't walk back from the head block repeatedly by failing if a duplicate block is requested:
```suggestion
// Ensure we don't walk back from head on every lookup by failing if a block is loaded multiple times.
	requestedBlocks := make(map[common.Hash]bool)
	blockByHash := func(hash common.Hash) *types.Block {
		if requestedBlocks[hash] {
			t.Fatalf("Requested duplicate block: %v", hash)
		}
		requestedBlocks[hash] = true
		return oracle.BlockByHash(hash, chainCfg.ChainID.Uint64())
	}
```